applications/vanilla/controllers/api/DraftsApiController.php

```diff
@@ -0,0 +1,394 @@
+ */
+class DraftsApiController extends AbstractApiController {
+
+    /** @var CapitalCaseScheme */
+    private $caseScheme;
+
+    /** @var DraftModel */
+    private $draftModel;
+
+    /** @var Schema */
+    private $idParamSchema;
+
+    /** @var CamelCaseScheme */
+    private $translateCaseScheme;
+
+    /**
+     * DraftsApiController constructor.
+     *
+     * @param DraftModel $draftModel
+     */
+    public function __construct(DraftModel $draftModel) {
+        $this->draftModel = $draftModel;
+        $this->caseScheme = new CapitalCaseScheme();
+        $this->translateCaseScheme = new CamelCaseScheme();
+    }
+
+    /**
+     * Delete a draft.
+     *
+     * @param int $id The unique ID of the draft.
+     */
+    public function delete($id) {
+        $this->permission('Garden.SignIn.Allow');
+
+        $in = $this->idParamSchema('in')->setDescription('Delete a draft.');
+        $out = $this->schema([], 'out');
+
+        $row = $this->draftByID($id);
+        if ($row['InsertUserID'] !== $this->getSession()->UserID) {
+            $this->permission('Garden.Moderation.Manage');
+        }
+        $this->draftModel->deleteID($id);
+    }
+
+    /**
+     * Get a draft by its unique ID.
+     *
+     * @param int $id
+     * @throws
+     * @return array
+     */
+    public function draftByID($id) {
+        $row = $this->draftModel->getID($id, DATASET_TYPE_ARRAY);
+        if (!$row) {
+            throw new NotFoundException('Draft');
+        }
+        return $row;
+    }
+
+    /**
+     * Get a draft schema with minimal add/edit fields.
+     *
+     * @param string $type The type of schema.
+     * @return Schema Returns a schema object.
+     */
+    public function draftPostSchema($type) {
+        static $draftPostSchema;
+
+        if (!isset($draftPostSchema)) {
+            $draftPostSchema = $this->schema(
+                Schema::parse(
+                    ['recordType', 'parentRecordID?', 'attributes']
+                )->add($this->fullSchema()),
+                'DraftPost'
+            );
+        }
+
+        return $this->schema($draftPostSchema, $type);
+    }
+
+    /**
+     * Get a schema instance comprised of all available draft fields.
+     *
+     * @return Schema Returns a schema object.
+     */
+    protected function fullSchema() {
+        static $schema;
+
+        if (!isset($schema)) {
+            $schema = Schema::parse([
+                'draftID:i' => 'The unique ID of the draft.',
+                'recordType:s' => [
+                    'description' => 'The type of record associated with this draft.',
+                    'enum' => ['comment', 'discussion']
+                ],
+                'parentRecordID:i|n' => 'The unique ID of the intended parent to this record.',
+                'attributes:o' => [
+                    'description' => 'A free-form object containing all custom data for this draft.',
+                    'properties' => [
```

Comment:
> The description and the validation should be done using the recordType schema

I don't think that's the route to go either. We can't dynamically change the schema based on the parameters it receives. It's possible a drafts UI might want to stash some information that isn't necessarily a column in the resource's table (a temporary selection of some UI element, etc.). Maybe there are fields we don't want to endorse the usage of in drafts by making them valid attributes for a resource. It's going to get increasingly trickier to make this flexible enough to be a valid for anything, but still including resource-specific validation. I think the simplest route is letting the UI stash whatever it wants.

I agree that the attributes here were too specific. They work great for now, but documenting these doesn't make much sense if the drafts API is going to be more open-ended.

Reply:
The description should say to look for the recordType model.
The validation should be done using the recordType schema but any value could be null.
Wouldn't that be the best solution?

Reply:
[One of the requirements mentioned on the original task](https://github.com/vanilla/vanilla/issues/5535#issuecomment-330909057) was for the attributes field to "load/save anything given to the draft". I interpreted this to mean *anything*. I am concerned that enforcing additional (and dynamic) validation requirements might work well enough for Vanilla's current capabilities with discussions and comments, but may limit the usefulness for drafts of other resource types in the future.

Reply:
If a recordType is specified we should validate that each field is either null or matching the schema field's type and any extra fields should be left untouched. This would meet all the requirements and give some insight of what should be excepted by the API user.

This could be done in a future iteration though. We could just not validate the attribute data at all for now.